pkg/server/server.go

```diff
@@ -499,6 +503,7 @@ func (s *ProxyServer) serveRecvBackend(stream agent.AgentService_ConnectServer,
 				client.agentID = agentID
 				s.addFrontend(agentID, resp.ConnectID, client)
 				close(client.connected)
+				metrics.Metrics.ObserveDialLatency(time.Since(client.start))
```

Comment:
Looks like we fail to record latency on error. Is that desired? Also should be be recording a dial error metric?

Reply:
+1. Similar to how dial failures are recorded in the egress_selector in k/k, it might be helpful to also record dial failures here

Reply:
Recording dial failures here does not add much to the existing k/k dial error metrics. If we want to know how many dial has failed, we can check the k/k metrics. If we want to know the reason of dial failure, we can check the error message in k/k.

Recording the latency is different, it helps break down the dial latency. Though in my test it show the dial latency isn't a performance bottleneck.

Reply:
Not recording latency on error is intentional. This is to avoid failures like dial timeout from bloating the average latency. Also this is consistent with the egressselector [metrics](https://github.com/kubernetes/kubernetes/blob/6079cebfae672335e958e03e7b0c258ba264d28f/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go#L225-L239) in k/k.